Fixes #28315 - allow autocomplete without form scope

ezr-ondrej opened the pull request:
This is follow up for #6980 where we forgot that `auto_complete_search` is part of an API for plugins and should have some deprecation period.
This deprecates the method and allows following:

```erb
<%= form_with url: '/hola_amigos' do |f| %>
  <%= autocomplete_f(f, :search, value: 'default') %>
<% end %>
```
What should allow the transfer path for autocompletes, which needs `name="search"` without scope.

ezr-ondrej commented:
@sharvit could you take a look as you were the one merging #6980 where I have caused this?

ezr-ondrej committed:
Fixes #28315 - allow autocomplete without form scope

adamruzicka commented:
Needs a rebase

adamruzicka commented:
This allows REX to work again even without https://github.com/theforeman/foreman_remote_execution/pull/449, however it is still kind-of broken, see the image in the REX PR.

ezr-ondrej commented:
This has become some nasty craftsmanship :(
I've had to extract the input component, because we have apparently need for solely the inputs without all the form input goods here and there.
Edit: dropped `ControlContext` as it made no sense :thinking:

sharvit commented on webpack/assets/javascripts/react_app/components/common/forms/InputFactory.js:
```diff
@@ -0,0 +1,56 @@
+  inputComponents[name] = Component;
+};
+
+const InputFactory = ({ type, ...controlProps }) => {
```
Can you please add unit-testing?

sharvit replied:
Not a blocker but how does it align with @xprazak2 workaround migrating to `formik`?

sharvit reviewed (commented):
Thanks for the fast response @ezr-ondrej :+1: 

@adamruzicka can you verify the latest changes fixes the issue with rex?

adamruzicka commented:
With this PR and the REX one it is still tiny.

With this PR and without the REX one it works fine, although the clear cross is a bit oddly placed.
![image](https://user-images.githubusercontent.com/7326770/69355173-a4989900-0c81-11ea-8a99-f84b4ec7483a.png)

sharvit commented:
@ezr-ondrej what is the status here?

ezr-ondrej commented:
Sorry, I were trying to figure out the tests. Finally I noticed the tests of the `FormInput`s are partially testing `InputFactory` so I moved the factory related to separated file. Is that ok, @sharvit?

adamruzicka commented:
What is the plan here?

With REX PR only the search box is tiny, with this PR only and with this PR together with the REX one it mostly works, but the clear cross is still oddly placed.

sharvit commented on webpack/assets/javascripts/react_app/components/common/forms/__tests__/InputFactory.test.js:
```diff
@@ -0,0 +1,45 @@
+  name: null,
+};
+
+describe('InputFactory', () => {
```
The testing approach makes sense to me.

Could you make it simpler by using the `testComponentSnapshotsWithFixtures` for the own component test?
Maybe by doing:
```js
  describe('rendering', () => {
    registerInputComponent('ownInput', Abc);
    testComponentSnapshotsWithFixtures(InputFactory, fixtures);
  });
```

sharvit reviewed (changes requested):
> Sorry, I were trying to figure out the tests. Finally I noticed the tests of the `FormInput`s are partially testing `InputFactory` so I moved the factory related to separated file. Is that ok, @sharvit?

Yes, it does make sense, thanks!


> With REX PR only the search box is tiny, with this PR only and with this PR together with the REX one it mostly works, but the clear cross is still oddly placed.

@ezr-ondrej do you have any idea about how to handle it?
I assume it can be fixed by a css change or by fixing a markup issue in REX

ezr-ondrej commented:
I've rebased on webcomponents and updated the tests.

> @ezr-ondrej do you have any idea about how to handle it?

This should be done, now together it behaves correctly, we didn't use the new option on the REX side.

adamruzicka commented:
What is the status here?

sharvit reviewed (changes requested):
Thanks @ezr-ondrej, I tested it together with https://github.com/theforeman/foreman/pull/6832 and I noticed snapshot changes when running npm test.

Looks like the `clear` icon is not aligned now, does it related to the `formik` change?
![Architectures](https://user-images.githubusercontent.com/1262502/70546723-79b2ae00-1b78-11ea-9e41-2894ea0a3e28.png)

ezr-ondrej commented:
> Looks like the `clear` icon is not aligned now, does it related to the `formik` change?

I guess it's not related, it just wants to go for a walk it seems. Have to look into the css for it :slightly_smiling_face:

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SearchBar/search-bar.scss:
```diff
@@ -1,15 +1,6 @@
 
 .search-bar {
-  .autocomplete-aux {
-    right: 108px;
```
This was a hack. I have made the field relative and we are positioning against it, so we do not need to count with the buttons and the css is ok everywhere.

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/AutoComplete/auto-complete.scss:
```diff
@@ -1,9 +1,11 @@
   .autocomplete-aux {
     position: absolute;
-    right: 25px;
+    right: 4px;
```
With positioning we can state what we want. - 4px from the right side.
:stop_sign: :man_cartwheeling: -> :walking_man:

ezr-ondrej commented:
I've fixed the css so I believe this is ready? :slightly_smiling_face:

sharvit commented:
Thanks @ezr-ondrej, does the Jenkins error related?

sharvit reviewed (approved):
Thanks @ezr-ondrej, looks good and works as expected :+1:

sharvit commented:
> Thanks @ezr-ondrej, does the Jenkins error related?

Not related

Ron-Lavi commented:
could you fix the storybook?
I am getting the following warnings:
```bash
WARNING in ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
78:0-22 "export 'registerInputComponent' was not found in './FormField'
 @ ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
 @ ./webpack/assets/javascripts/react_app .stories.js$
 @ ./webpack/stories/index.js
 @ ./.storybook/config.js
 @ multi ./node_modules/@storybook/react/dist/server/config/polyfills.js ./node_modules/@storybook/react/dist/server/config/globals.js (webpack)-hot-middleware/client.js?reload=true ./.storybook/config.js

WARNING in ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
81:32-46 "export 'ControlContext' was not found in './FormField'
 @ ./webpack/assets/javascripts/react_app/components/common/forms/Form.stories.js
 @ ./webpack/assets/javascripts/react_app .stories.js$
 @ ./webpack/stories/index.js
 @ ./.storybook/config.js
 @ multi ./node_modules/@storybook/react/dist/server/config/polyfills.js ./node_modules/@storybook/react/dist/server/config/globals.js (webpack)-hot-middleware/client.js?reload=true ./.storybook/config.js
```
The storybook is down right now but you can still see the issue, this PR will make it work soon: #7258

ezr-ondrej commented:
Thanks @laviro for noticing, I've opened #7263 to fix it.
